### ISSUE TITLE
Invalid inner header id size

### ISSUE THREAD

**Issue by Hadevmin** [keepassxreboot/keepassxc#8482]
Hi,

I'm really sorry but I need help. 
I have deleted a .kdbx file and unfortunately the bin has been emptied. 
When I realized this, I stopped using my SSD and used a tool to find a trace of my deleted file.

However when I open it I get the following message:
"Error while reading the database: Invalid inner header id size"

I tried to use the keepass repair function but without success.
What possibilities can I explore to try to get my keepassxc file open?

Thanks in advance for your help

**Comment by droidmonkey** [comment]
You have lost your database forever. Unfortunately there is absolutely nothing we or you can do at this point. You didn't have a backup?

**Comment by Hadevmin** [comment]
Hi @droidmonkey,
I have a backup but I am missing a very important password that was added between my backup and now.
Even if I go to a data recovery company there is no chance that the database will be readable again?

**Comment by droidmonkey** [comment]
No chance at all. SSD often wipe rows at random and data is generally stored in a pseudo random fashion across the NAND modules. If your initial recovery failed then there is absolutely nothing you can do to get your database back into a usable form.

**Comment by Hadevmin** [comment]
Okay, thanks for taking the time to respond. 
Great job! Keep up the good work!

**Comment by melroy89** [comment]
You know.. it's strange that this happened today on my PC and on **another** PC, with two different databases. Around the same time, again using different machines and hardware.

Both using the same KeePassXC version 2.7.4.

Same error message on both databases on different machines: `Error while reading the database: Invalid inner header id size`.

**Comment by droidmonkey** [comment]
Basically this error message means an attempt was made to read 1 byte of data from the inner header, but no bytes were actually read. This could happen if the file was truncated at this point in the file or the inner header was incompletely written (very unlikely).

**Comment by melroy89** [comment]
Thanks for for additional info.

I will run a **memtest** on my server. I'm using Nextcloud to sync the file to storage, so if my PC dies, I won't lose my database. But maybe my server is misbehaving.. 

Since this issue happened on both machines (on different computers), but using the same Nextcloud instance. I actually suspect the server at this point. Maybe the file then got synced back from the server to the local machine, who knows.. I'm kinda panicking now, since I don't know what to trust. It could be RAM, it could be LVM. It could be anything at this point.

**Comment by droidmonkey** [comment]
Yah it wasn't clear in your message, it is the same database synced across two different machines.

**Comment by melroy89** [comment]
> Yah it wasn't clear in your message, it is the same database synced across two different machines.

No actually. Two different databases on two different machines,. Both got corrupted at roughly the same time. Luckily I have back-ups and also Nextcloud makes "versions".

**Comment by droidmonkey** [comment]
Yah sounds like something broke on the server side with sync then. The probability of that error happening independently is zero.

**Comment by melroy89** [comment]
Update: Same database file that was "corrupted" yesterday, was tried today again (so not the backup). This time no error message in KeePassXC desktop.. (nor on mobile app)? 

It automatically _uncorrupted_ itself.. Impossible.. 

EDIT: Maybe Jupiter or Saturn was in the wrong place at the wrong time.. What is happening here.. I actually tried to set the date time back to yesterday on my machine (2023-03-31 23:10:00) trying to replicate the error, but I was unable to.

**Comment by RichardoC** [comment]
For anyone seeing this, I've had this when I've used rclone mount to mount the remote folder containing my DB locally. 

Stopping and starting the mount fixed this issue, it actually seemed to be a problem where the file being presented to keepassxc was corrupted, but the file on the server was fine. 

Below is the relevant rclone log line (verbose logging)
```console
2023/05/02 19:00:51 ERROR : IO error: corrupted on transfer: md5 hash differ "72c36606d50da11ad8cd231f531a5bba" vs "7f239b55be76959941d44a0f5f74a3ab"

2023/05/02 19:00:51 ERROR : MyVault.kdbx: ReadFileHandle.Release error: corrupted on transfer: md5 hash differ "72c36606d50da11ad8cd231f531a5bba" vs "7f239b55be76959941d44a0f5f74a3ab"
```

**Comment by melroy89** [comment]
Interesting. Luckily you have a (remote) backup in this case. Of course everybody should have backups.

I'm actually more interested personally why it went corrupt. The logging says corrupted on transfer. But who saids it was not corrupted locally and not because of a network transfer.

Anyway, getting the db file corrupted is very hard to replicate consistently. Sure, files can get corrupted by many reasons. 

In the past 25 years I never met software that causes so many corruptions in a single file. I really think it deserves a thorough investigation in the KeePass software. Despite this issue being "closed", but obviously not really resolved.

**Comment by droidmonkey** [comment]
> But who saids it was not corrupted locally and not because of a network transfer.

If that were the case then rclone would have no idea and transfer the corrupted file without issue. The error here occurred outside of keepassxc (it would seem to me).

What could possibly cause this issue is the special way we save the file in order to prevent half-saves. We use QSaveFile that writes a temp file in the same directory, deletes the original, and atomically moves this temp file into place. Sync programs can get screwed up when this happens.

**Comment by melroy89** [comment]
> > But who saids it was not corrupted locally and not because of a network transfer.
> 
> If that were the case then rclone would have no idea and transfer the corrupted file without issue. The error here occurred outside of keepassxc (it would seem to me).
> 
> What could possibly cause this issue is the special way we save the file in order to prevent half-saves. We use QSaveFile that writes a temp file in the same directory, deletes the original, and atomically moves this temp file into place. Sync programs can get screwed up when this happens.

Interesting. I'm using nextcloud. Which is also a sync program but won't use rsync. I think?

Anyway, what you describe does make sense. And is generally a good way to do this. Maybe things go wrong if the file was locked or something.

**Comment by andpol** [comment]
To help future readers, I came across this same issue today with keepassxc + gdrive rclone mount. The problem in my case was the rclone vfs cache (which I have set to 999 days), but for some reason had a corrupted copy of the database file. The file was fine in Google Drive. Wiping the local cache (defaulted to `~/.cache/rclone` here) fixed all my issues, and no data loss.

**Comment by RedSnt** [comment]
Glad Onedrive keeps copies of different versions of the same file - even though it was most likely also the problem.  
Added an account+pw to my DB using KeePassDX on android, and opening the DB on my PC it had corrupted.  
I've changed the settings in KeePassDX and KeePassXC to **stop auto-saving** - pretty sure that's what messing with onedrive's active syncronization.

**Comment by mboltonjr** [comment]
I experienced the same problem when moving documents from a Windows machine to Linux.  Copied files via a USB stick and then moved them to the Linux docs directory.  Opened the KeepassXC file and got the error.  

I later mounted the network drives via Samba and copied the same DB file to the local docs directory, and it suddenly opened.

**Comment by droidmonkey** [comment]
USB keys are notorious for corrupting files, especially when trying to do a quick copy job like that. Good chance data was still being written to the drive when it was pulled. Always transfer data via network of you can, lots of safeguard built in to prevent these issues.

**Comment by phoerious** [comment]
If you use USB drives, at least eject them safely, unplug, then plug back in and check whether everything is still readable and complete. If you see errors there, dispose of the device. Don't try to repair it.

**Comment by mboltonjr** [comment]
Drive is always ejected safely and checked periodically with professional disk utilities.  So that's not an issue, especially when all the other files copied were fine.  I believe the Linux version of KeePass was doing something a little different compared to the Windows portable edition.

COR 251753

**Comment by phoerious** [comment]
No, the two versions do exactly the same. Other files being intact or the health metrics reported by some "professional" disk utility are no indicators of a fully functioning flash device.

**Comment by mboltonjr** [comment]
Well, regardless, it worked.  Didn't matter what was used or how it got there.  It worked.  And I do have 10 backups of the same file on different disks in the Windows Data Center array.  

Oh and for the record, if there's already bits of a file missing on a 2 year old WD Purple MicroSD card, I'm not confident in relying on an SSD or NVMe drive for storage or an OS.  My "PROFESSIONAL" disk diagnostics (SpinRite) tool has witnessed severe degrading of cells in a particular section of an SSD memory chip (various mftrs).   Good for eternal reads but still significantly bad for writes.

WD VelociRaptures and WDs newer Gold series drives continue to spin along after 7 years or constant read/writes.  And they to this day have no bad sectors or warning signs.  That's 65,829 hrs on the OS drive.

Anyway, it's time I stick to using a laptop HD as a file transfer disk here on.  Let that be a warning for others who are dependent on USB/MicroSD drives.

FNL 260037

**Comment by Antik9421** [comment]
Unfortunately, it's unrecoverable. My only hope is to find backups.
I encountered a situation where my Google Drive file became corrupted. I tried everything and was about to give up. The AI ​​suggested the photorec tool to find the deleted temporary file. I created the signature photorec.sig.
```
kdbx 0 0x03d9a29a67fb4bb5
```

and it found about 20 files, each several GB in size. Fortunately, some of them opened, and after resaving, they were now 200-300 KB in size.
